packages/bigquery/src/index.js

```diff
@@ -57,16 +57,30 @@ var Table = require('./table.js');
  *
  * @resource [What is BigQuery?]{@link https://cloud.google.com/bigquery/what-is-bigquery}
  *
+ * The apiEndpoint from options will set the host. If not set, the
+ * `GOOGLE_CLOUD_BIGQUERY_ENDPOINT` environment variable is honored,
+ * otherwise the actual API endpoint will be used.
+ *
  * @param {object} options - [Configuration object](#/docs).
+ * @param {object} options.apiEndpoint - Override the default API endpoint
+ * to reach BigQuery.
  */
 function BigQuery(options) {
   if (!(this instanceof BigQuery)) {
     options = common.util.normalizeArguments(this, options);
     return new BigQuery(options);
   }
 
+  var basePath = '/bigquery/v2';
+
+  var baseInfo = common.util.determineBaseUrl(
+      options,
+      'GOOGLE_CLOUD_BIGQUERY_ENDPOINT',
+      'https://www.googleapis.com');
+
   var config = {
-    baseUrl: 'https://www.googleapis.com/bigquery/v2',
+    baseUrl: baseInfo.apiEndpoint + basePath,
+    customEndpoint: baseInfo.customEndpoint,
```

Comment:
In the case of services extending common.Service it should work fine. I can think of two edge cases though:

- We cannot use the same baseUrl for all the protoServices, for example bigtable has different endpoints for the actual service vs. the admin service. At the moment this is taken care of in the bigtable constructor.
- For backward compatibility reasons (not sure how important it's for you guys) I kept support for [SERVICE]_EMULATOR_HOST when it was already used. Bringing the baseUrl logic down to the services means that we'll have to support an array of environment variables, it's doable but feels like a hack rather than a clean solution